# Capstone Modelling

Building off the previous notebook of preprocessing and data development.

## Preprocessing and Training Data Development

This notebook outlines the steps taken to preprocess the data and prepare it for model training based on the 'metdata_cleaned.csv' dataset.

### Step 1: Loading the Dataset

First, we load the cleaned dataset into a DataFrame.

In [1]:
import pandas as pd

# Load the cleaned data
df = pd.read_csv('metdata_cleaned.csv')

df.head()

objectID  isHighlight accessionNumber  accessionYear  isPublicDomain  \
0         1        False      1979.486.1         1979.0           False   
1         2        False      1980.264.5         1980.0           False   
2         3        False        67.265.9         1967.0           False   
3         4        False       67.265.10         1967.0           False   
4         5        False       67.265.11         1967.0           False   

  primaryImage primaryImageSmall additionalImages  \
0          NaN               NaN               []   
1          NaN               NaN               []   
2          NaN               NaN               []   
3          NaN               NaN               []   
4          NaN               NaN               []   

                                        constituents         department  ...  \
0  [{'constituentID': 164292, 'role': 'Maker', 'n...  The American Wing  ...   
1  [{'constituentID': 1079, 'role': 'Maker', 'nam...  The American Wing  ...   
2                                                NaN  The American Wing  ...   
3                                                NaN  The American Wing  ...   
4                                                NaN  The American Wing  ...   

  classification rightsAndReproduction linkResource              metadataDate  \
0            NaN                   NaN          NaN  2021-04-06T04:41:04.967Z   
1            NaN                   NaN          NaN  2021-04-06T04:41:04.967Z   
2            NaN                   NaN          NaN  2021-04-06T04:41:04.967Z   
3            NaN                   NaN          NaN  2024-01-10T04:57:19.843Z   
4            NaN                   NaN          NaN  2024-01-10T04:57:19.843Z   

                                 repository  \
0  Metropolitan Museum of Art, New York, NY   
1  Metropolitan Museum of Art, New York, NY   
2  Metropolitan Museum of Art, New York, NY   
3  Metropolitan Museum of Art, New York, NY   
4  Metropolitan Museum of Art, New York, NY   

                                           objectURL  tags objectWikidata_URL  \
0  https://www.metmuseum.org/art/collection/search/1   NaN                NaN   
1  https://www.metmuseum.org/art/collection/search/2   NaN                NaN   
2  https://www.metmuseum.org/art/collection/search/3   NaN                NaN   
3  https://www.metmuseum.org/art/collection/search/4   NaN                NaN   
4  https://www.metmuseum.org/art/collection/search/5   NaN                NaN   

  isTimelineWork GalleryNumber  
0          False           NaN  
1          False           NaN  
2          False           NaN  
3          False           NaN  
4          False           NaN  

[5 rows x 57 columns]

### Step 2: Creating Dummy Variables

We identify and process categorical variables such as 'department', 'objectName', and 'culture' for dummy encoding.

In [2]:
# Creating dummy variables for selected categorical variables
categorical_vars = ['department', 'objectName', 'culture']
df_processed = pd.get_dummies(df, columns=categorical_vars, drop_first=True)

df_processed.head()

objectID  isHighlight accessionNumber  accessionYear  isPublicDomain  \
0         1        False      1979.486.1         1979.0           False   
1         2        False      1980.264.5         1980.0           False   
2         3        False        67.265.9         1967.0           False   
3         4        False       67.265.10         1967.0           False   
4         5        False       67.265.11         1967.0           False   

  primaryImage primaryImageSmall additionalImages  \
0          NaN               NaN               []   
1          NaN               NaN               []   
2          NaN               NaN               []   
3          NaN               NaN               []   
4          NaN               NaN               []   

                                        constituents  \
0  [{'constituentID': 164292, 'role': 'Maker', 'n...   
1  [{'constituentID': 1079, 'role': 'Maker', 'nam...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                          title  ...  culture_Chinese, for American market  \
0  One-dollar Liberty Head Coin  ...                                     0   
1  Ten-dollar Liberty Head Coin  ...                                     0   
2    Two-and-a-Half Dollar Coin  ...                                     0   
3    Two-and-a-Half Dollar Coin  ...                                     0   
4    Two-and-a-Half Dollar Coin  ...                                     0   

   culture_Dutch  culture_Dutch, probably  culture_European culture_French  \
0              0                        0                 0              0   
1              0                        0                 0              0   
2              0                        0                 0              0   
3              0                        0                 0              0   
4              0                        0                 0              0   

  culture_French, possibly culture_German culture_Guatemalan culture_Mexican  \
0                        0              0                  0               0   
1                        0              0                  0               0   
2                        0              0                  0               0   
3                        0              0                  0               0   
4                        0              0                  0               0   

  culture_Spanish  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 194 columns]

### Step 3: Standardizing Numeric Features

We standardize numeric features to have a mean of 0 and a standard deviation of 1. Here, we'll standardize 'accessionYear', 'objectBeginDate', and 'objectEndDate' as examples.

In [3]:
from sklearn.preprocessing import StandardScaler

# Selecting numeric columns for standardization
numeric_columns = ['accessionYear', 'objectBeginDate', 'objectEndDate']
scaler = StandardScaler()
df_processed[numeric_columns] = scaler.fit_transform(df_processed[numeric_columns])

df_processed.head()

objectID  isHighlight accessionNumber  accessionYear  isPublicDomain  \
0         1        False      1979.486.1       1.516240           False   
1         2        False      1980.264.5       1.554907           False   
2         3        False        67.265.9       1.052244           False   
3         4        False       67.265.10       1.052244           False   
4         5        False       67.265.11       1.052244           False   

  primaryImage primaryImageSmall additionalImages  \
0          NaN               NaN               []   
1          NaN               NaN               []   
2          NaN               NaN               []   
3          NaN               NaN               []   
4          NaN               NaN               []   

                                        constituents  \
0  [{'constituentID': 164292, 'role': 'Maker', 'n...   
1  [{'constituentID': 1079, 'role': 'Maker', 'nam...   
2                                                NaN   
3                                                NaN   
4                                                NaN   

                          title  ...  culture_Chinese, for American market  \
0  One-dollar Liberty Head Coin  ...                                     0   
1  Ten-dollar Liberty Head Coin  ...                                     0   
2    Two-and-a-Half Dollar Coin  ...                                     0   
3    Two-and-a-Half Dollar Coin  ...                                     0   
4    Two-and-a-Half Dollar Coin  ...                                     0   

   culture_Dutch  culture_Dutch, probably  culture_European culture_French  \
0              0                        0                 0              0   
1              0                        0                 0              0   
2              0                        0                 0              0   
3              0                        0                 0              0   
4              0                        0                 0              0   

  culture_French, possibly culture_German culture_Guatemalan culture_Mexican  \
0                        0              0                  0               0   
1                        0              0                  0               0   
2                        0              0                  0               0   
3                        0              0                  0               0   
4                        0              0                  0               0   

  culture_Spanish  
0               0  
1               0  
2               0  
3               0  
4               0  

[5 rows x 194 columns]

### Step 4: Splitting the Data

We split the dataset into training and testing sets. Note: The target variable needs to be defined based on your model's goal.

In [4]:
from sklearn.model_selection import train_test_split

# Placeholder for feature selection
# You'll need to adjust 'features' and 'target' to your specific case
features = df_processed.drop('accessionYear', axis=1)
target = df_processed['accessionYear']

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)

print(f'Training set size: {X_train.shape[0]}')
print(f'Testing set size: {X_test.shape[0]}')


Training set size: 800
Testing set size: 200


## Model Fitting

In this project, we worked on a dataset ('metdata_cleaned.csv') containing information about artifacts from the Metropolitan Museum of Art. The goal was to build a classification model to predict whether an artifact is in the public domain or not.

### Data Preprocessing:

We started by loading the dataset using pandas and created dummy variables for categorical features 'department' and 'culture'.
The target variable 'isPublicDomain' was separated from the feature dataset.
To handle missing values, we performed imputation using the median strategy on numeric columns to ensure robustness against outliers.
Initially, we encountered a mismatch in the number of columns between the imputed data and the feature dataset, which was resolved by selecting only the first 7 numeric columns for imputation.

### Model Training and Evaluation:

After preprocessing, we standardized the feature dataset using StandardScaler to ensure all features have a mean of 0 and a standard deviation of 1.
The dataset was split into training and testing sets with a ratio of 80:20 using train_test_split.
We trained a Logistic Regression model with a maximum iteration of 1000 on the training data.
Model evaluation was performed on the test set using various metrics such as accuracy, precision, recall, F1 score, and ROC-AUC score.
The performance metrics of the trained logistic regression model were printed, showing an accuracy of 0.84, precision of 0.83, recall of 0.99, F1 score of 0.90, and ROC-AUC score of 0.80.


### Example Model 1: Decision Tree

In [34]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer

# Load the dataset
file_path = 'metdata_cleaned.csv'
df = pd.read_csv(file_path)

# Creating dummy variables for 'department' and 'culture'
df = pd.get_dummies(df, columns=['department', 'culture'], drop_first=True)

# Prepare the target variable
y = df['isPublicDomain']

# Exclude the non-numeric 'accessionNumber' column and any other non-numeric columns
df_features = df.drop(['isPublicDomain', 'accessionNumber'], axis=1)

# Select only numeric columns for imputation to avoid the error
numeric_columns = df_features.select_dtypes(include=['int64', 'float64']).columns

# Apply imputation to numeric columns only
imputer = SimpleImputer(strategy='median')
imputed_data = imputer.fit_transform(df_features[numeric_columns])

Shape of imputed_data: (1000, 7)
Columns in numeric_columns: Index(['objectID', 'accessionYear', 'period', 'dynasty', 'reign', 'portfolio',
       'artistBeginDate', 'artistEndDate', 'objectBeginDate', 'objectEndDate',
       'subregion', 'locale', 'locus', 'excavation', 'river', 'classification',
       'linkResource', 'GalleryNumber'],
      dtype='object')


In [35]:
# Adjust numeric_columns to contain only the columns present in imputed_data
# Grab only the first 7 columns of numeric_columns
numeric_columns = df_features.select_dtypes(include=['int64', 'float64']).columns[:7]

# Convert the imputed data back to a DataFrame
imputed_df = pd.DataFrame(imputed_data, columns=numeric_columns)

# Update the df_features with imputed data
df_features[numeric_columns] = imputed_df


In [38]:
from sklearn.impute import SimpleImputer

# Impute missing values with mean
imputer = SimpleImputer(strategy='mean')
X_imputed = imputer.fit_transform(X_scaled)

# Continue with train-test split, model training, and evaluation
X_train, X_test, y_train, y_test = train_test_split(X_imputed, y, test_size=0.2, random_state=42)
lr_model = LogisticRegression(max_iter=1000)
lr_model.fit(X_train, y_train)
y_pred = lr_model.predict(X_test)
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, lr_model.predict_proba(X_test)[:, 1])

print(f'Accuracy: {accuracy}\nPrecision: {precision}\nRecall: {recall}\nF1 Score: {f1}\nROC-AUC Score: {roc_auc}')


Accuracy: 0.84
Precision: 0.8305084745762712
Recall: 0.9865771812080537
F1 Score: 0.901840490797546
ROC-AUC Score: 0.8049743387287802


### Random Forest

In [43]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score
from sklearn.preprocessing import StandardScaler
from sklearn.impute import SimpleImputer

# Load the dataset
file_path = 'metdata_cleaned.csv'
df = pd.read_csv(file_path)

# Creating dummy variables for 'department' and 'culture'
df = pd.get_dummies(df, columns=['department', 'culture'], drop_first=True)

# Prepare the target variable
y = df['isPublicDomain']

# Exclude non-numeric columns
numeric_columns = df.select_dtypes(include=['int64', 'float64']).columns

# Prepare features
X = df[numeric_columns]

# Impute missing values
imputer = SimpleImputer(strategy='median')
X_imputed = imputer.fit_transform(X)

# Standardize the features
scaler = StandardScaler()
X_scaled = scaler.fit_transform(X_imputed)

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.2, random_state=42)

# Initialize the Random Forest classifier
rf_model = RandomForestClassifier(n_estimators=100, random_state=42)

# Train the Random Forest model
rf_model.fit(X_train, y_train)

# Predict on the test set
y_pred = rf_model.predict(X_test)

# Calculate evaluation metrics
accuracy = accuracy_score(y_test, y_pred)
precision = precision_score(y_test, y_pred)
recall = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)
roc_auc = roc_auc_score(y_test, rf_model.predict_proba(X_test)[:, 1])

# Print the evaluation metrics
print(f'Random Forest Model Performance:')
print(f'Accuracy: {accuracy}\nPrecision: {precision}\nRecall: {recall}\nF1 Score: {f1}\nROC-AUC Score: {roc_auc}')


Random Forest Model Performance:
Accuracy: 0.92
Precision: 0.9290322580645162
Recall: 0.9664429530201343
F1 Score: 0.9473684210526316
ROC-AUC Score: 0.9416370575075668


Comparing the performance metrics of the logistic regression model and the random forest model:

1. **Accuracy:**
   - Logistic Regression: 0.84
   - Random Forest: 0.92
   
   The random forest model has a higher accuracy compared to the logistic regression model.

2. **Precision:**
   - Logistic Regression: 0.8305
   - Random Forest: 0.929
   
   The random forest model also has a higher precision, indicating a better ability to correctly classify positive predictions.

3. **Recall:**
   - Logistic Regression: 0.9866
   - Random Forest: 0.9664
   
   The logistic regression model has a higher recall, suggesting it is better at capturing true positives.

4. **F1 Score:**
   - Logistic Regression: 0.9018
   - Random Forest: 0.9474
   
   The random forest model has a higher F1 score, which balances precision and recall.

5. **ROC-AUC Score:**
   - Logistic Regression: 0.805
   - Random Forest: 0.942
   
   The random forest model has a significantly higher ROC-AUC score, indicating better overall performance in distinguishing between the two classes.

In summary, based on these performance metrics, the random forest model generally outperforms the logistic regression model for this classification task. However, the choice between models also depends on the specific requirements and constraints of the project, such as computational resources, interpretability, and ease of implementation.